# make database and annotate lung-related genes in different species
``` shell
miniprot -in genomes/chicken.fa -dbtype nucl -out genomes/chicken.fa
...
miniprot -t 32 --outc 0.8 --gff genomes/chicken.fa.idx allgene.pep > minimap_out/chicken.gff
...
```
# Find genes near CNE
```python
#########Find_neighbor_gene.py######################
#!/usr/bin/env python
import re, os
from collections import defaultdict

WINDOW_SIZE = 2000000

GENE_LIST = 'gene.lst'
GFF_DIR = 'minimap_out'
BLAST_DIR = 'blast_out_filter'
OUTFILE = 'shared.cnes.2m.plus.distance'

def parse_gene_miniprot(file, other_gene_info, gene_dict):
    match = re.search(r'([\w_]+).gff$', file)
    if not match: raise ValueError(f"Filename {file} does not match expected format.")
    species = match.group(1)
    # print(species)
    if species not in gene_dict: gene_dict[species] = {}

    with open(file) as f:
        for line in f:
            if not line.startswith('##PAF'): continue
            data = re.split(r'\s+', line.strip())
            
            common_id, chromosome, start, end, strand = data[1], data[6], int(data[8]), int(data[9]), data[5]
            match = re.search(r';([^;]+)$', common_id)
            if not match: raise ValueError(f"common_id {common_id} does not match expected format.")
            common_id = match.group(1)
            flag = f"{chromosome}:{int(start) // WINDOW_SIZE * WINDOW_SIZE}"
            
            if flag not in gene_dict[species]: 
                gene_dict[species][flag] = set()
            gene_dict[species][flag].add(common_id)
            
            if common_id not in other_gene_info: 
                other_gene_info[common_id] = {}
            
            if species not in other_gene_info[common_id]:
                other_gene_info[common_id][species] = {flag: (start, end, strand)}
            else:
                flag_data = other_gene_info[common_id][species]
                if flag in flag_data:
                    old_start, old_end, old_strand = flag_data[flag]
                    new_start = min(start, old_start)
                    new_end = max(end, old_end)
                    flag_data[flag] = (new_start, new_end, old_strand)
                else:
                    flag_data[flag] = (start, end, strand)

def process_blast(file, other_gene_info, gene_dict, result, repeat, total_aln_count):
    species = file.split('/')[-1].split('.')[0]
    with open(file, 'r') as f:
        for line in f:
            values = re.split(r'\s+', line.strip())
            if len(values) < 12: continue
            query_id, ref_id, ref_start, ref_end = values[0], values[1], int(values[8]), int(values[9])
            cne_id = re.match(r".*:\w+-\w+:(cne\d+)", query_id).group(1)
            if cne_id not in total_aln_count:
                total_aln_count[cne_id] = set()
            total_aln_count[cne_id].add(species)
            repeat[cne_id] += 1

            flags = [
                f"{ref_id}:{(ref_start // WINDOW_SIZE) * WINDOW_SIZE}",
                f"{ref_id}:{(ref_start // WINDOW_SIZE - 1) * WINDOW_SIZE}",
                f"{ref_id}:{(ref_start // WINDOW_SIZE + 1) * WINDOW_SIZE}"
            ]

            for flag in flags:
                if flag in gene_dict[species]:
                    for common_id in gene_dict[species][flag]:
                        for old_flag in other_gene_info[common_id][species]:
                            start, end, strand = other_gene_info[common_id][species][old_flag]
                            if end + WINDOW_SIZE > ref_start and start - WINDOW_SIZE < ref_end:
                                distance = 0
                                if ref_start > end: 
                                    distance = ref_start -end
                                if ref_end < start:
                                    distance = ref_end - start
                                if strand == '-':
                                    distance = distance * -1
                                result[common_id][cne_id][species] = distance

def main():
    result, repeat = defaultdict(lambda: defaultdict(dict)), defaultdict(int)
    
    other_gene_info = gene_dict = {}

    gff_files = [f for f in os.listdir(GFF_DIR) if f.endswith('.gff')]
    for file in gff_files:
        gff_file = os.path.join(GFF_DIR, file)
        parse_gene_miniprot(gff_file, other_gene_info, gene_dict)

    total_aln_count = {}
    blast_files = [f for f in os.listdir(BLAST_DIR) if f.endswith('.out')]
    for file in blast_files:
        blast_file = os.path.join(BLAST_DIR, file)
        process_blast(blast_file, other_gene_info, gene_dict, result, repeat, total_aln_count)

    with open(OUTFILE, 'w') as output:
        output.write(f'gene_id\tcne_id\tmatched_species_count\tpair_count\tdetailed_info\n')
        for common_id, cnes in result.items():
            for cne_id, species in cnes.items():
                formatted_species = [f'{species}:{distance}' for species, distance in species.items()]
                content = ' '.join(sorted(formatted_species))
                count = len(species)
                matched_species_count = len(total_aln_count[cne_id])  # 新增：计算比对上的物种数量
                output.write(f'{common_id}\t{cne_id}\t{matched_species_count}\t{count}\t{content}\n')

if __name__ == '__main__':
    main()

```